heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,541 @@
+    }
+  }
+
+  /**
+   * Schedule a new task
+   *
+   * @param taskId the taskId of container to handle
+   */
+  protected boolean scheduleNewTask(String taskId) {
+    // Put it to the pending start queue
+    LOG.info(String.format("We are to schedule task: [%s]", taskId));
+
+    // Update the tasksId
+    int containerIndex = TaskUtils.getContainerIndexForTaskId(taskId);
+    tasksId.put(containerIndex, taskId);
+
+    // Re-schedule it
+    toScheduleTasks.add(taskId);
+    LOG.info(String.format("Added task: %s into the to-schedule-tasks queue: ", taskId));
+    return true;
+  }
+
+  @Override
+  public void frameworkMessage(SchedulerDriver schedulerDriver,
+                               Protos.ExecutorID executorID,
+                               Protos.SlaveID slaveID,
+                               byte[] data) {
+    // TODO(mfu): TO handle this message
+    LOG.info("Received framework message.");
+  }
+
+  @Override
+  public void disconnected(SchedulerDriver schedulerDriver) {
+    // TODO(mfu): TO handle this failure
+    LOG.info("Disconnected!");
+  }
+
+  @Override
+  public void slaveLost(SchedulerDriver schedulerDriver, Protos.SlaveID slaveID) {
+    // TODO(mfu): TO handle this failure
+    LOG.info("Slave lost: " + slaveID.toString());
+  }
+
+  @Override
+  public void executorLost(SchedulerDriver schedulerDriver,
+                           Protos.ExecutorID executorID,
+                           Protos.SlaveID slaveID,
+                           int status) {
+    // TODO(mfu): TO handle this failure
+    LOG.info(String.format("Executor lost: %s, slave ID: %s, status: %d",
+        executorID.toString(), slaveID.toString(), status));
+  }
+
+  @Override
+  public void error(SchedulerDriver schedulerDriver, String message) {
+    // TODO(mfu): TO handle this failure
+    LOG.info("Received error: " + message);
+  }
+
+
+  /**
+   * Generate launchable tasks basing on offer resources
+   *
+   * @param offerResources Mapping from Protos.Offer to utils class TaskResources
+   * @return list of launchable task
+   */
+  protected List<LaunchableTask> generateLaunchableTasks(
+      Map<Protos.Offer, TaskResources> offerResources) {
+    List<LaunchableTask> tasks = new LinkedList<>();
+
+    if (isTerminated) {
+      LOG.info("Job has been killed");
+      return tasks;
+    }
+
+    while (!toScheduleTasks.isEmpty()) {
+      String taskId = toScheduleTasks.poll();
+      BaseContainer baseContainer =
+          containersInfo.get(TaskUtils.getContainerIndexForTaskId(taskId));
+
+      TaskResources neededResources =
+          new TaskResources(
+              baseContainer.cpu,
+              baseContainer.memInMB,
+              baseContainer.diskInMB,
+              baseContainer.ports);
+
+      Iterator<Map.Entry<Protos.Offer, TaskResources>> it = offerResources.entrySet()
+          .iterator();
+      while (it.hasNext()) {
```

Comment:
The resource assignment strategy here is "first-fit", which may not be a very good strategy in practise. Should we abstract this assignment logic here out and refine it further?